### PR TITLE
feat: Chat Message History.

### PR DESCRIPTION
**Pull request by JU-2094**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [X] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/langchain-google-datastore-python/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [X] Ensure the tests and linter pass
- [X] Code coverage does not decrease (if any source code was changed)
- [X] Appropriate docs were updated (if necessary)

Add Chat Message History for Firestore in Datastore Mode.

### CODE REVIEW

File: src/langchain_google_datastore/chat_message_history.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        """Chat Message History for Google Cloud Firestore.
+        Args:
+            session_id: Arbitrary key that is used to store the messages of a single
+                chat session. This is the identifier of an entity.
+            kind: The name of the Datastore kind to write into.
+            client: Client for interacting with the Google Cloud Firestore API.
+        """
+        if client:
+            self.client = client
+            self.client._client_info.user_agent = USER_AGENT
```
Comment (kurtisvg) [inline src/langchain_google_datastore/chat_message_history.py:47]: this is overriding the agent entirely. Can we preserve the original as well?

@averikitsch

File: src/langchain_google_datastore/chat_message_history.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        else:
+            self.client = datastore.Client()
+            self.client._client_info.user_agent = USER_AGENT
```
Comment (kurtisvg) [inline src/langchain_google_datastore/chat_message_history.py:50]: nit: simplify
```suggestion
        if not client:
            self.client = datastore.Client()
        self.client._client_info.user_agent = USER_AGENT

```

File: src/tests/unit_tests/test_chat_message_history.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+
+def test_firestore_load_messages(test_case: TestCase) -> None:
+    NUM_MESSAGES = 25
+    session_id = uuid.uuid4().hex
+    chat_history = DatastoreChatMessageHistory(
+        session_id=session_id, kind="HistoryLoad"
+    )
+
+    expected_messages: List[AIMessage | HumanMessage] = []
```
Comment (kurtisvg) [inline src/tests/unit_tests/test_chat_message_history.py:61]: LangChain has a BaseMessage class

**Review by kurtisvg — APPROVED** [review]

File: src/langchain_google_datastore/chat_message_history.py (hunk 1 above)
Comment (JU-2094) [reply src/langchain_google_datastore/chat_message_history.py:47]: I'm not very familiar with the USER_AGENT_HEADER .

Should we add an space and the agent followed by the version?   
I'm looking at some examples like
```
Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Mobile Safari/537.36
```

I'm not certain about this. I can append it to the string if it's better

File: src/langchain_google_datastore/chat_message_history.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        self,
+        session_id: str,
+        kind: str = DEFAULT_KIND,
+        client: Optional[Client] = None,
+    ) -> None:
+        """Chat Message History for Google Cloud Firestore.
+        Args:
+            session_id: Arbitrary key that is used to store the messages of a single
+                chat session. This is the identifier of an entity.
+            kind: The name of the Datastore kind to write into.
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_datastore/chat_message_history.py:42]: DONE

File: src/langchain_google_datastore/chat_message_history.py (hunk 2 above)
Comment (JU-2094) [reply src/langchain_google_datastore/chat_message_history.py:50]: DONE

File: src/tests/unit_tests/test_chat_message_history.py (hunk 3 above)
Comment (JU-2094) [reply src/tests/unit_tests/test_chat_message_history.py]: DONE

File: src/langchain_google_datastore/chat_message_history.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        entity = self.client.entity(self.key)
+        entity["messages"] = MessageConverter.encode_messages(self.messages)
+        self.client.put(entity)
+
+    def clear(self) -> None:
+        self.messages = []
+        self.client.delete(self.key)
+
+
+class MessageConverter:
```
[the remark this answers is not in the document]
Comment (JU-2094) [reply src/langchain_google_datastore/chat_message_history.py:76]: Not really, I was using the utility class pattern - (more like java)
I don't know if the best practice is to have this as functions of the file. 

Should I change it?

**Pull request merged**